pennylane/ops/functions/equal.py

```diff
@@ -0,0 +1,97 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the qml.equal function.
+"""
+# pylint: disable=too-many-arguments,too-many-return-statements
+import pennylane as qml
+
+
+def equal(op1, op2, check_interface=True, check_trainability=True, rtol=1e-5, atol=1e-9):
+    r"""Function for determining operator equality.
+
+    Args:
+        op1 (.Operator): First operator to compare
+        op2 (.Operator): Second operator to compare
+        check_interface (bool, optional): Whether to compare interfaces. Default: `True`
+        check_trainability (bool, optional): Whether to compare trainability status. Default: `True`
+        rtol (float, optional): Relative tolerance for parameters
+        atol (float, optional): Absolute tolerance for parameters
+
+    Returns:
+        bool: `True` if the operators are equal, else `False`
+
+    **Example**
+
+    Given two operators, ``qml.equal`` determines their equality:
+
+    >>> op1 = qml.RX(np.array(.12), wires=0)
+    >>> op2 = qml.RY(np.array(1.23), wires=0)
+    >>> qml.equal(op1, op1), qml.equal(op1, op2)
+    True False
+
+    .. details::
+        :title: Usage Details
+
+        You can use the optional arguments to get more specific results.
+
+        Consider the following comparisons:
+
+        >>> op1 = qml.RX(torch.tensor(1.2), wires=0)
+        >>> op2 = qml.RX(jax.numpy.array(1.2), wires=0)
+        >>> qml.equal(op1, op2)
+        False
+
+        >>> qml.equal(op1, op2, check_interface=False, check_trainability=False)
+        True
+
+        >>> op3 = qml.RX(np.array(1.2, requires_grad=True), wires=0)
+        >>> op4 = qml.RX(np.array(1.2, requires_grad=False), wires=0)
+        >>> qml.equal(op3, op4)
+        False
+
+        >>> qml.equal(op3, op4, check_trainability=False)
+        True
+
+        If the parameters do not have a `requires_grad` attribute, the function will return `False` if `check_trainability`
+        is `True`.
+    """
+    if op1.__class__ is not op2.__class__:
+        return False
+    if not all(
+        qml.math.allclose(d1, d2, rtol=rtol, atol=atol) for d1, d2 in zip(op1.data, op2.data)
+    ):
+        return False
+    if not op1.wires == op2.wires:
+        return False
+    for kwarg in op1.hyperparameters:
+        if not op1.hyperparameters[kwarg] == op2.hyperparameters[kwarg]:
+            return False
+
+    if check_trainability:
+        for params_1, params_2 in zip(op1.data, op2.data):
+            # Vanilla NumPy and JAX do not have the `requires_grad` attribute.
+            # This will raise an `AttributeError` for such cases.
+            try:
+                if not params_1.requires_grad == params_2.requires_grad:
```

Comment:
Could probably use [`qml.math.requires_grad`](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.math.requires_grad.html) for this instead of having to use a try-except block.  This function would also catch the trainability of TensorFlow variables, and it would allow the use of scalars, lists, and vanilla numpy.

Reply:
Ah yes! I wasn't aware of this function. Makes much more sense.